### DISCUSSION TITLE
UMAT-DLL load problem

### DISCUSSION THREAD

**Discussion by Geotechnique-simlation** [Anura3D/Anura3D_OpenSource#14] [Q&A]
Dear developer, I have the following questions, please give me some advice. "DLL can not be loaded, how can we solve the problem?"

**Comment by murphy-john** [comment]
You can try looking [here](https://github.com/Anura3D/Anura3D_OpenSource/discussions/3#discussioncomment-2892820).

Please review the existing discussions before posting issues or questions.

**Reply by Geotechnique-simlation** [reply]
Well, I have reviewed the existing discussions, but still haven't solved my problem. The following figure is a problem when I run：
![QQ截图20220615085049](https://user-images.githubusercontent.com/107266329/173713517-51b1f2d1-cdc4-47a3-8df4-1e099479b62c.png)

**Reply by murphy-john** [reply]
I'm not familiar with the "SandModify" constitutive model. Are you running a tutorial example or your own model?  This error is caused by the code not seeing the DLL file in the folder you're running the model in. In this case, it doesn't see a DLL called "SandModify" in the "Column collapse.A3D" folder.

If you're running a user defined soil model that's external to the models included with the source code, you'll have to make sure they are compatible with Anura3D. You can do this using the debug feature in Visual Studio.

**Reply by Geotechnique-simlation** [reply]
The "Sandmodify" model is my own model. The generated DLL file is generated according to the user manual tutorial. Even if I copy the "Sandmodify" model to the "column collapse.a3d" folder, the above error will still occur

**Reply by murphy-john** [reply]
I see, I'm not too familiar with external user models. Where in the user manual tutorial are you seeing the instructions on creating and using external DLLs?

I looked at the code briefly. One thing that might help you is opening the GOM file and change the name of the $$MATERIAL_MODEL_SOLID from "user_defined_soil_model" to "external_soil_model".

**Reply by FrancescaCeccato** [reply]
it seems that the code find the dll, but cannot load it. This means that there must be an error int  he interface between the code and Anura. I see that you are using Anura3D.2019.2, did you prepare the dll following the instructions of the tutaorial manual 2019 or 2020? we changed some options and this migh be the problem.
I suggest to move to version 2022.

**Reply by Geotechnique-simlation** [reply]
Dear Collaborators, there is a bug with the "WaveSpeed is not a number or infinite trying to write results" when I utilized the external soil model.

**Reply by murphy-john** [reply]
Will you describe the bug with the Wavespeed calculation you're seeing with the external soil model? Or are you getting the wave speed error when you do your calculation? 

Have you tried debugging the code to find the issue?

**Reply by pmarveggio** [reply]
The "WaveSpeed is not a number or infinite trying to write results" error message you obtained may be related to a missing definition of the stiffness matrix in the UMAT. Did you initialize its value?
